### PR TITLE
GDV-13: [C++] Add support for filters

### PR DESCRIPTION
**Pull request by pravindra**
- similar to projection, filter is built for a specific schema and condition (i.e expression)
- the output of filter is a selection vector (Int16Array or int32Array)

### CODE REVIEW

File: cpp/src/codegen/tree_expr_builder.cc [hunk 1]
```diff
@@ -148,4 +148,24 @@ ExpressionPtr TreeExprBuilder::MakeExpression(const std::string &function,
...
+    std::cout << "Condition: " << root_node->ToString() << "\n";
+  }
+
+  return ConditionPtr(new Condition(root_node));
+}
+
+ConditionPtr TreeExprBuilder::MakeCondition(const std::string &function,
+                                            const FieldVector &in_fields) {
+  std::vector<NodePtr> field_nodes;
+  for (auto it = in_fields.begin(); it != in_fields.end(); ++it) {
```
Comment (praveenbingo) [inline cpp/src/codegen/tree_expr_builder.cc]: for each for consistency..

File: cpp/src/codegen/tree_expr_builder.cc [hunk 2]
```diff
@@ -148,4 +148,24 @@ ExpressionPtr TreeExprBuilder::MakeExpression(const std::string &function,
   return MakeExpression(func_node, out_field);
 }
 
+ConditionPtr TreeExprBuilder::MakeCondition(NodePtr root_node) {
```
Comment (praveenbingo) [inline cpp/src/codegen/tree_expr_builder.cc]: please add a null check..

File: cpp/include/gandiva/condition.h [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+#include "gandiva/expression.h"
+#include "gandiva/gandiva_aliases.h"
+
+namespace gandiva {
+
+/// \brief A condition expression.
+class Condition : public Expression {
+ public:
+  Condition(const NodePtr root)
+      : Expression(root, std::make_shared<arrow::Field>("cond", arrow::boolean())) {}
```
Comment (praveenbingo) [inline cpp/include/gandiva/condition.h]: should we validate that return type is boolean in the expression validator?

Comment (pravindra) [reply cpp/include/gandiva/condition.h]: this already happens in the expression validator.

https://github.com/dremio/gandiva/blob/master/cpp/src/codegen/expr_validator.cc#L35

File: cpp/src/codegen/tree_expr_builder.cc (hunk 2 above)
Comment (pravindra) [reply cpp/src/codegen/tree_expr_builder.cc]: will do,

File: cpp/src/codegen/tree_expr_builder.cc (hunk 1 above)
Comment (pravindra) [reply cpp/src/codegen/tree_expr_builder.cc]: will do.

**Pull request merged**